### DISCUSSION TITLE
Method enabled by __WASM__ in a cross runtime library for WASM isn't getting picked up in a WASM project

### DISCUSSION THREAD

**Discussion by roxk** [unoplatform/uno#10015] [Q&A]
Steps to reproduce:
1. Create a new cross runtime library as documented [here](https://platform.uno/docs/articles/get-started-dotnet-new.html?tabs=windows#uno-platform-blank-cross-runtime-library)
2. In the generated WASM project, add a method surrounded by `__WASM__`.
3. Create a new uno WASM project and consume the library as nuget
4. Observe the method added in (2) isn't available.

**Comment by ivanjx** [comment]
please post your csproj files

**Reply by jeromelaban** [reply]
Indeed, it will be best if you post a sample repro project which shows what you're trying to do. Thanks!

**Reply by roxk** [reply]
I think my repro steps are quite clear, did you try it out? Anyways, I can create a minimal repro.

**Reply by FrozDark** [reply]
We need your repro sample

**Reply by roxk** [reply]
https://github.com/roxk/UnoCrossRuntimeWasmNotWorking

Here you are, the readme contains more info. Thanks!

**Comment by ivanjx** [comment]
try to add `__WASM__` to your `ConsumingApp.Wasm.csproj`'s `DefineConstants`

**Reply by roxk** [reply]
It doesn't sound right. In the minimal repro there is already a `__WASM__` guard and that is why these lines are being compiled but failed anyways:

https://github.com/roxk/UnoCrossRuntimeWasmNotWorking/blob/main/ConsumingApp/ConsumingApp/ConsumingApp.Shared/MainPage.xaml.cs#L33

or, as shown in this screenshot, these lines are syntax highlighted so the constant is already defined:

![image](https://user-images.githubusercontent.com/16918354/193102731-d2176dac-02d8-4847-8f7e-188262190413.png)

As a more technical answer, the WASM head (created with uno's template...) references `Uno.UI.WebAssembly` which reference `Uno.UI.Runtime.WebAssembly` which already defines the constant.

**Reply by roxk** [reply]
Put it the other way, is there anything else a WASM head needs to do to consume cross-runtime library to correctly pick up the `uno-runtime/webassembly/$namespace.Wasm.dll`?

**Reply by jeromelaban** [reply]
This is an expected behavior. Cross-runtime libraries are only exposing what's in the "reference" API (`CrossRuntimeLib/CrossRuntimeLib.csproj` in your solution). This means that any project referencing this nuget package will not get access to those platform specific APIs (Skia or Wasm).

One way to handle this is to do something like this:
```csharp
public void WasmOnlyMethod()
{
#if __WASM__
   // wasm specific code
#else
   throw new PlatformNotSupportedException();
#endif
}
```

**Reply by roxk** [reply]
This is not suitable for my use case as the `WasmOnlyMethod` needs parameter and that parameter is a uno only type.

Other cross-target libraries do not have this limitation and can add whatever method + types however they like in MonoAndroid or iOS. For example, adding a method that accepts `Context` in android and `UIApplication` in iOS. Why would uno have this limitation? What other solutions do I have now? This is library-breaking.

**Reply by jeromelaban** [reply]
The technique to handle this scenarion is with two packages, where those packages are built in the same solution. 

For instance, you can make cross-runtime `PackageA` (your current package), and a second WebAssembly-only package named `PackageA.WebAssembly`. In that platform specific package, you can use WebAssembly-specific features that can be exposed to WebAssembly heads. You can create extension methods, or other helpers to get access to platform specific features, with platform specific types.

Uno uses this technique with the `Uno.UI.Runtime.WebAssembly` to expose WebAssembly only features to WebAssembly apps, like [HtmlElementAttribute](https://platform.uno/docs/articles/interop/wasm-javascript-1.html#uno-wasm-controls-are-actual-html5-elements).

For the reason this is needed, the problem here is that Skia and WebAssembly are using the same `TargetFramework`, and there's no way for current .NET tooling, _**in a single package**_, to discriminate like you would with iOS (`net6.0-ios`), Android (`net6.0-android`), MacCatalyst, Windows, etc... which all have their own `TargetFrameworkIdentifier`.

This is a .NET and NuGet limitation, but also a feature which is also known as "bait and switch" materialized as `RuntimeIdentifier` switching.

**Reply by roxk** [reply]
In my case I need the WASM method to be part of a partial class and thus having the method defined in another package doesn't work. Extension method doesn't work as the method I'm trying to define is a constructor.

This situation is a bit like how I needed to access the `UIElement(string)` ctor defined in `UIElement.wasm.cs` but it isn't available in a WASM project. Same issue, manifested in two difference scenarios.

> For the reason this is needed

I thought there was some build time logic employed by uno to pick up the WASM binary at build time but maybe this is not feasible. Thanks.

Uno itself just works, but TBF this library/multi-targeting situation isn't great. Sounds like a good reason to propose a wasm TFM, or dotnet will need a more sophisticated/flexible mechanism to pick binary.

In the mean time, I will bite the bullet and make a "reference API" accepting an `object`, and inside the method try to cast the object to the correct type (in a `__WASM__` guard).

**Reply by jeromelaban** [reply]
> In the meantime, I will bite the bullet and make a "reference API" accepting an `object`, and inside the method try to cast the object to the correct type (in a `__WASM__` guard).

If you need to provide an instance through the constructor, in the `MyControl.WebAssembly` package, create a helper class with a method that will `new` your class with the proper parameters `MyControl MyControlHelper.Create(MyWasmCustomParameter) { ... }`.

> I thought there was some build time logic employed by uno to pick up the WASM binary at build time but maybe this is not feasible. Thanks.

It's not the current behavior as while it may work for heads, it will not work for libraries which only have one target framework available (`netstandard2.0`) which covers both Wasm and Skia based targets.

> Uno itself just works, but TBF this library/multi-targeting situation isn't great. Sounds like a good reason to propose a wasm TFM, or dotnet will need a more sophisticated/flexible mechanism to pick binary.

It may be, yet WebAssembly is not just one environment (the browser) and locking onto one particular environment at this time is premature, as there's already newer environment appearing like wasi. For now, a common API with RID replacements is good enough, and is forward compatible.

**Comment by roxk** [comment]
@jeromelaban I'm getting a namespace not resolved error. In the updated minimal repro which now compiles, I'm getting `.nuget\packages\uno.wasm.bootstrap\3.3.1\build\Uno.Wasm.Bootstrap.targets(175,5): error : [00:00:00.5895220] Unhandled exception. System.Exception: Could not resolve CrossRuntimeLib` in the cosuming app project. Anything I missed?

(The doc certainly could use some update)

**Reply by jeromelaban** [reply]
We're taking contributions for documentation updates, if you're up for it and have found improvements to make.

This particular error message is not enough to troubleshoot the error. If you can provide a repro solution, we'll be able to look at the same issue.

**Reply by roxk** [reply]
The repro is the same minimal repro above. In case you missed it: https://github.com/roxk/UnoCrossRuntimeWasmNotWorking

**Reply by jeromelaban** [reply]
Thanks. Package references for local projects are not well supported, you'll need to use a project reference instead. I'm guessing you built the package once and got it installed through a local nuget source?

**Reply by roxk** [reply]
For my case that library is another package available on nuget. I'm adding Uno support for it so making it a project reference is not possible. Installing through a local nuget source is just for minimal repro purposes.

**Reply by jeromelaban** [reply]
I see. For building purposes, as a matter of being able to test the library in the solution that builds it, you'll still need to use `ProjectReference`.

So there's indeed an issue in the original template, where the assembly for the WebAssembly project should be the same as the "reference" one. So adding `<AssemblyName>CrossRuntimeLib</AssemblyName>` in the `CrossRuntimeLib.Wasm` project will help with the error message you got. We'll fix the template.

**Reply by roxk** [reply]
Thanks! I will test this and report if it works.

_Update_: it works in the minimal repro. Will test in the real app.

**Reply by roxk** [reply]
In the real app, I get `packages\uno.wasm.bootstrap\3.3.1\build\Uno.Wasm.Bootstrap.targets(175,5): error : System.Exception: Failed to execute the linker`. Will try to repro in the minimal repro. Is referencing `Xamarin.Essentials` in the `CrossRuntimeLib` project, and referencing `Uno.Xamarin.Essentials` in the `CrossRuntimeLib.Wasm` project a possible source of error?

**Reply by jeromelaban** [reply]
If you can provide the [whole build log](https://platform.uno/docs/articles/uno-builds-troubleshooting.html), it'll be easier to troubleshoot.

**Reply by roxk** [reply]
@jeromelaban The problem was caused by dependency indeed, and I have fixed that (removing any package reference related to `XamarinEssentials` in WASM project and move it to the netstandard project fixed it), and the wasm project can now consume the wasm-aware cross runtime lib and compiles. I'm now facing another issue: At launch, console log shows `$namespace.GlobalStaticResources` couldn't be found. I could reproduce it in the [same minimal repro](https://github.com/roxk/UnoCrossRuntimeWasmNotWorking), please take a look.

Enclosed is a screenshot of the content of the two binaries of the lib, as shown via dotpeek:

![image](https://user-images.githubusercontent.com/16918354/194251878-f471617c-ba55-4a83-ac4a-fd0d14abd8ae.png)

You can see `GlobalStaticResources` is indeed not present in the WASM binary, whereas it's present in the ordinary netstandard binary. 

Thanks!

**Reply by roxk** [reply]
Since nuget isn't working, I'm trying to copy all source of the library (all code is in a shared project anyways) to the consuming app to consuming it. Now the same `GlobalStaticResources` missing error is thrown on `Uno.UI.Wasm` by `Uno.Xamarin.Essentials`...If it helps, my project is a Uno UWP -> WASM project, _not_ a uno xamarin.form -> WASM project. Version of `Uno.Xamarin.Essentials` is latest, i.e. `1.4.1-uno.13`.

![image](https://user-images.githubusercontent.com/16918354/194260770-987fcc2f-91ad-491a-93a2-1c098fc8f214.png)

**Reply by jeromelaban** [reply]
Thanks for the update. This likely means that the Uno.Xamarin.Essentials package is having an issue with the runtime replacement and the end of the build, which is likely because it was not updated to support this feature. 

We have not pursued updating Uno.Xamarin.Essentials because most developers are using the WinRT APIs directly, instead of going through Xamarin.Essentials.

**Reply by roxk** [reply]
This breaks any library which was using xamarin essentials though. It's fine, I can remove the reference to xamarin essential for when `__WASM__` is present in the library, but it'd be great if uno can update that forked package. Xamarin.Essentials won't get any new features anyways.
